arch: arc: fix a reg misuse in leaving tickless idle

YuguoWH opened the pull request:
There is a register misuse in leaving tickless idle code, which would destroy exception/interrupt status which is saved in r0. 
This PR fix this issue by using r3 instead of r0

Fixes #29220

abrodkin reviewed (approved):
That's a good one!
Though why do we need `exit_tickless_idle` macro if it is only used once?
Maybe opencode this all right in place?

ruuddw reviewed (changes requested):
Agree that it would be better to not have a macro but put this inline.
If kept as a macro, please add a comment what registers are used (and fix the comment on line 282 that only r0 is stomped -> r3 as well.

ruuddw reviewed (approved):
Thanks for the clean-up. Just one minor last thing: the comments now talk about r0 and r3, but also r1 is used. As it is not a macro, I'm ok with simply removing the info on what registers are used (we don't have that consistently for other assembler either).

YuguoWH commented:
Hi @ruuddw , I'll keep it simple here. Thanks for advices.

abrodkin commented on arch/arc/core/isr_wrapper.S:
```diff
@@ -279,8 +262,21 @@ SECTION_FUNC(TEXT, _isr_demux)
 	/* cannot be done before this point because we must be able to run C */
-	/* r0 is available to be stomped here, and exit_tickless_idle uses it */
-	exit_tickless_idle
+	/* exit tickless idle */
```
Sorry for that last minute request, but can we move `/* exit tickless idle */` comment under `#ifdef CONFIG_SYS_POWER_MANAGEMENT` and also switch from `#if defined` to more usual `#ifdef`?

YuguoWH replied:
I found there're several places that use `#ifdef` in isr_wrapper.s, so I put the changes in a seperate commit.

abrodkin commented:
Nothing else to add here - well done :)
I guess it's OK to merge now.

evgeniy-paltsev commented on arch/arc/core/isr_wrapper.S:
```diff
@@ -279,8 +262,21 @@ SECTION_FUNC(TEXT, _isr_demux)
-	/* r0 is available to be stomped here, and exit_tickless_idle uses it */
-	exit_tickless_idle
+#if defined(CONFIG_SYS_POWER_MANAGEMENT)
+	/* exit tickless idle */
```
Can we keep `exit_tickless_idle` macro instead of moving its code to _isr_demux?
I'd prefer to keep this logical code separation to make it more understandable.

YuguoWH replied:
This part of code only used once, therefore I think it is better to not seperate it as a macro.

evgeniy-paltsev replied:
No mater it's used only one - this code part has specific functionality so lets keep it logically separated and named with macro to avoid our ASM functions to become endless monolithic piece of code.

YuguoWH replied:
Modified. Thanks for advice.

YuguoWH committed:
arch: arc: fix a reg misuse in leaving tickless idle

There is a register misuse in leaving tickless idle code, which would
destroy exception/interrupt status. This commit fix this issue.

Signed-off-by: Yuguo Zou <yuguo.zou@synopsys.com>

YuguoWH committed:
arch: arc: use ifdef to replace if define in isr wrapper code

isr wrapper code has mixed usage of #ifdef and #if define macros. Unify
them to more usual #ifdef.

Signed-off-by: Yuguo Zou <yuguo.zou@synopsys.com>
